src/chainstate/burn/operations/peg_in.rs

```diff
@@ -100,18 +108,18 @@ impl PegInOp {
 
         let standard_principal_data = StandardPrincipalData(version, address_data);
 
-        let memo = data.get(61..).unwrap_or(&[]).to_vec();
+        let (recipient, bytes_read) =
+            if let Some((contract_name, bytes_read)) = Self::read_contract_name(&data[21..])? {
+                let contract_id =
+                    QualifiedContractIdentifier::new(standard_principal_data, contract_name);
 
-        let recipient: PrincipalData =
-            if let Some(contract_bytes) = Self::leading_non_zero_bytes(data, 21, 61) {
-                let contract_name: String = std::str::from_utf8(contract_bytes)?.to_owned();
-
-                QualifiedContractIdentifier::new(standard_principal_data, contract_name.try_into()?)
-                    .into()
+                (contract_id.into(), 21 + bytes_read)
             } else {
-                standard_principal_data.into()
+                (standard_principal_data.into(), 22)
             };
 
+        let memo = data.get(bytes_read..).unwrap_or(&[]).to_vec();
```

Comment:
I don't think we support variable-sized memos. The memo should only be the space unused by the rest of the fields.